### ISSUE TITLE
Driver Installation Cancelled by User

### ISSUE THREAD

**Issue by jrhug** [pbatard/libwdi#209]
I originally installed this driver and then had to remove it.  Now, I am trying to reinstall it and I get the warning:
**libwdi:warning [CreateSelfSignedCert] Could not obtain a key container: The requested key container is corrupted.**  

Where do I find this "corrupted key container"?  I was successful in loading the driver once.  Do I need to take extra actions when uninstalling?  The system security settings did not change between the original installation and the attempt to reinstall. 

**Log**
```
Using inf name: Flyswatter2_(Interface_1).inf
Successfully extracted driver files.
Installing driver. Please wait...
libwdi:info [extract_binaries] successfully extracted driver files to C:\Users\hughjame\usb_driver
libwdi:info [wdi_prepare_driver] successfully created 'C:\Users\hughjame\usb_driver\Flyswatter2_(Interface_1).inf'
libwdi:info [wdi_prepare_driver] Creating and self-signing a .cat file...
libwdi:debug [AddFileHash] 'wdfcoinstaller01011.dll': PE type
libwdi:info [ScanDirAndHash] added hash for 'C:\Users\hughjame\usb_driver\amd64\wdfcoinstaller01011.dll'
libwdi:debug [AddFileHash] 'winusbcoinstaller2.dll': PE type
libwdi:info [ScanDirAndHash] added hash for 'C:\Users\hughjame\usb_driver\amd64\winusbcoinstaller2.dll'
libwdi:debug [AddFileHash] 'flyswatter2_(interface_1).inf': INF type
libwdi:info [ScanDirAndHash] added hash for 'C:\Users\hughjame\usb_driver\flyswatter2_(interface_1).inf'
libwdi:debug [AddFileHash] 'wdfcoinstaller01011.dll': PE type
libwdi:info [ScanDirAndHash] added hash for 'C:\Users\hughjame\usb_driver\x86\wdfcoinstaller01011.dll'
libwdi:debug [AddFileHash] 'winusbcoinstaller2.dll': PE type
libwdi:info [ScanDirAndHash] added hash for 'C:\Users\hughjame\usb_driver\x86\winusbcoinstaller2.dll'
libwdi:info [CreateCat] Successfully created file 'C:\Users\hughjame\usb_driver\Flyswatter2_(Interface_1).cat'
libwdi:debug [CreateSelfSignedCert] Set Enhanced Key Usage, URL and CPS
libwdi:warning [CreateSelfSignedCert] Could not obtain a key container: The requested key container is corrupted.
libwdi:warning [wdi_prepare_driver] could not sign cat file
libwdi:debug [wdi_install_driver] using progress bar mode
libwdi:debug [installer process] got parameter Flyswatter2_(Interface_1).inf
libwdi:debug [process_message] got request for device_id
libwdi:debug [installer process] got device_id: 'USB\VID_0403&PID_6010&MI_01\7&1CE5AEFC&0&0001'
libwdi:debug [process_message] got request for hardware_id
libwdi:debug [installer process] got hardware_id: 'USB\VID_0403&PID_6010&REV_0700&MI_01'
libwdi:debug [installer process] got user_sid: 'S-1-5-21-839522115-1383384898-515967899-5572085'
libwdi:debug [installer process] using syslog 'C:\WINDOWS\inf\setupapi.dev.log'
libwdi:debug [installer process] syslog reader thread started
libwdi:debug [installer process] successfully disabled the system restore point creation setting
libwdi:debug [process_message] switching timeout to infinite
libwdi:debug [installer process] Installing driver for USB\VID_0403&PID_6010&REV_0700&MI_01 - please wait...
libwdi:debug [syslog] <Garbled data>
libwdi:debug [syslog] >>>  [Device Install (UpdateDriverForPlugAndPlayDevices) - USB\VID_0403&PID_6010&REV_0700&MI_01]
libwdi:debug [syslog] >>>  Section start 2021/02/02 17:50:03.697
libwdi:debug [syslog]       cmd: "C:\Users\hughjame\usb_driver\installer_x64.exe" "Flyswatter2_(Interface_1).inf"
libwdi:debug [syslog]      ndv: INF path: C:\Users\hughjame\usb_driver\Flyswatter2_(Interface_1).inf
libwdi:debug [syslog]      ndv: Install flags: 0x00000001
libwdi:debug [syslog]      ndv: {Update Device Driver - USB\VID_0403&PID_6010&MI_01\7&1CE5AEFC&0&0001}
libwdi:debug [syslog]      ndv:      Search options: 0x00000080
libwdi:debug [syslog]      ndv:      Searching single INF 'C:\Users\hughjame\usb_driver\Flyswatter2_(Interface_1).inf'
libwdi:debug [syslog]      dvi:      {Build Driver List} 17:50:03.719
libwdi:debug [syslog]      dvi:           Searching for hardware ID(s):
libwdi:debug [syslog]      dvi:                usb\vid_0403&pid_6010&rev_0700&mi_01
libwdi:debug [syslog]      dvi:                usb\vid_0403&pid_6010&mi_01
libwdi:debug [syslog]      dvi:           Searching for compatible ID(s):
libwdi:debug [syslog]      dvi:                usb\class_ff&subclass_ff&prot_ff
libwdi:debug [syslog]      dvi:                usb\class_ff&subclass_ff
libwdi:debug [syslog]      dvi:                usb\class_ff
libwdi:debug [syslog]      sig:           {_VERIFY_FILE_SIGNATURE} 17:50:04.116
libwdi:debug [syslog]      sig:                Key      = flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                FilePath = c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                Catalog  = c:\users\hughjame\usb_driver\Flyswatter2_(Interface_1).cat
libwdi:debug [syslog] !    sig:                Verifying file against specific (valid) catalog failed.
libwdi:debug [syslog] !    sig:                Error 0x800b0100: No signature was present in the subject.
libwdi:debug [syslog]      sig:           {_VERIFY_FILE_SIGNATURE exit(0x800b0100)} 17:50:04.217
libwdi:debug [syslog]      sig:           {_VERIFY_FILE_SIGNATURE} 17:50:04.220
libwdi:debug [syslog]      sig:                Key      = flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                FilePath = c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                Catalog  = c:\users\hughjame\usb_driver\Flyswatter2_(Interface_1).cat
libwdi:debug [syslog] !    sig:                Verifying file against specific Authenticode(tm) catalog failed.
libwdi:debug [syslog] !    sig:                Error 0x800b0100: No signature was present in the subject.
libwdi:debug [syslog]      sig:           {_VERIFY_FILE_SIGNATURE exit(0x800b0100)} 17:50:04.251
libwdi:debug [syslog]      dvi:           Created Driver Node:
libwdi:debug [syslog]      dvi:                HardwareID   - USB\VID_0403&PID_6010&MI_01
libwdi:debug [syslog]      dvi:                InfName      - c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      dvi:                DevDesc      - Flyswatter2 (Interface 1)
libwdi:debug [syslog]      dvi:                Section      - USB_Install
libwdi:debug [syslog]      dvi:                Rank         - 0x80ff0001
libwdi:debug [syslog]      dvi:                Signer Score - Not digitally signed
libwdi:debug [syslog]      dvi:                DrvDate      - 08/09/2018
libwdi:debug [syslog]      dvi:                Version      - 6.1.7600.16385
libwdi:debug [syslog]      dvi:      {Build Driver List - exit(0x00000000)} 17:50:04.272
libwdi:debug [syslog]      dvi:      {DIF_SELECTBESTCOMPATDRV} 17:50:04.274
libwdi:debug [syslog]      dvi:           Default installer: Enter 17:50:04.299
libwdi:debug [syslog]      dvi:                {Select Best Driver}
libwdi:debug [syslog]      dvi:                     Class GUID of device changed to: {88bae032-5a81-49f0-bc3d-a4ff138216d6}.
libwdi:debug [syslog]      dvi:                     Selected Driver:
libwdi:debug [syslog]      dvi:                          Description - Flyswatter2 (Interface 1)
libwdi:debug [syslog]      dvi:                          InfFile     - c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      dvi:                          Section     - USB_Install
libwdi:debug [syslog]      dvi:                {Select Best Driver - exit(0x00000000)}
libwdi:debug [syslog]      dvi:           Default installer: Exit
libwdi:debug [syslog]      dvi:      {DIF_SELECTBESTCOMPATDRV - exit(0x00000000)} 17:50:04.305
libwdi:debug [syslog]      ndv:      Force Installing Driver:
libwdi:debug [syslog]      ndv:           Inf Name       - flyswatter2_(interface_1).inf
libwdi:debug [syslog]      ndv:           Driver Date    - 08/09/2018
libwdi:debug [syslog]      ndv:           Driver Version - 6.1.7600.16385
libwdi:debug [syslog]      sto:      {Setup Import Driver Package: c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf} 17:50:04.313
libwdi:debug [syslog]      inf:           Provider: libwdi
libwdi:debug [syslog]      inf:           Class GUID: {88bae032-5a81-49f0-bc3d-a4ff138216d6}
libwdi:debug [syslog]      inf:           Driver Version: 08/09/2018,6.1.7600.16385
libwdi:debug [syslog]      inf:           Catalog File: Flyswatter2_(Interface_1).cat
libwdi:debug [syslog]      sto:           {Copy Driver Package: c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf} 17:50:04.324
libwdi:debug [syslog]      sto:                Driver Package = c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sto:                Flags          = 0x00000007
libwdi:debug [syslog]      sto:                Destination    = C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}
libwdi:debug [syslog]      sto:                Copying driver package files to 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}'.
libwdi:debug [syslog]      flq:                Copying 'c:\users\hughjame\usb_driver\amd64\WdfCoInstaller01011.dll' to 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\amd64\WdfCoInstaller01011.dll'.
libwdi:debug [syslog]      flq:                Copying 'c:\users\hughjame\usb_driver\amd64\WinUSBCoInstaller2.dll' to 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\amd64\WinUSBCoInstaller2.dll'.
libwdi:debug [syslog]      flq:                Copying 'c:\users\hughjame\usb_driver\Flyswatter2_(Interface_1).cat' to 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\Flyswatter2_(Interface_1).cat'.
libwdi:debug [syslog]      flq:                Copying 'c:\users\hughjame\usb_driver\flyswatter2_(interface_1).inf' to 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\flyswatter2_(interface_1).inf'.
libwdi:debug [syslog]      sto:           {Copy Driver Package: exit(0x00000000)} 17:50:04.428
libwdi:debug [syslog]      pol:           {Driver package policy check} 17:50:04.565
libwdi:debug [syslog]      pol:           {Driver package policy check - exit(0x00000000)} 17:50:04.565
libwdi:debug [syslog]      sto:           {Stage Driver Package: C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\flyswatter2_(interface_1).inf} 17:50:04.566
libwdi:debug [syslog]      inf:                {Query Configurability: C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\flyswatter2_(interface_1).inf} 17:50:04.611
libwdi:debug [syslog]      inf:                     Driver package uses WDF.
libwdi:debug [syslog]      inf:                     Driver package 'flyswatter2_(interface_1).inf' is configurable.
libwdi:debug [syslog]      inf:                {Query Configurability: exit(0x00000000)} 17:50:04.618
libwdi:debug [syslog]      flq:                Copying 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\amd64\WdfCoInstaller01011.dll' to 'C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\amd64\WdfCoInstaller01011.dll'.
libwdi:debug [syslog]      flq:                Copying 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\amd64\WinUSBCoInstaller2.dll' to 'C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\amd64\WinUSBCoInstaller2.dll'.
libwdi:debug [syslog]      flq:                Copying 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\Flyswatter2_(Interface_1).cat' to 'C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\Flyswatter2_(Interface_1).cat'.
libwdi:debug [syslog]      flq:                Copying 'C:\Users\hughjame\AppData\Local\Temp\{a8983904-149a-f541-a27e-216ed470b1c6}\flyswatter2_(interface_1).inf' to 'C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\flyswatter2_(interface_1).inf'.
libwdi:debug [syslog]      sto:                {DRIVERSTORE IMPORT VALIDATE} 17:50:04.704
libwdi:debug [syslog]      sig:                     {_VERIFY_FILE_SIGNATURE} 17:50:04.761
libwdi:debug [syslog]      sig:                          Key      = flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                          FilePath = C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                          Catalog  = C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\Flyswatter2_(Interface_1).cat
libwdi:debug [syslog] !    sig:                          Verifying file against specific (valid) catalog failed.
libwdi:debug [syslog] !    sig:                          Error 0x800b0100: No signature was present in the subject.
libwdi:debug [syslog]      sig:                     {_VERIFY_FILE_SIGNATURE exit(0x800b0100)} 17:50:04.768
libwdi:debug [syslog]      sig:                     {_VERIFY_FILE_SIGNATURE} 17:50:04.771
libwdi:debug [syslog]      sig:                          Key      = flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                          FilePath = C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\flyswatter2_(interface_1).inf
libwdi:debug [syslog]      sig:                          Catalog  = C:\WINDOWS\System32\DriverStore\Temp\{ec0accc5-b311-3146-8a8a-8cbbcd2e5d40}\Flyswatter2_(Interface_1).cat
libwdi:debug [syslog] !    sig:                          Verifying file against specific Authenticode(tm) catalog failed.
libwdi:debug [syslog] !    sig:                          Error 0x800b0100: No signature was present in the subject.
libwdi:debug [syslog]      sig:                     {_VERIFY_FILE_SIGNATURE exit(0x800b0100)} 17:50:04.776
libwdi:debug [syslog] !!!  sig:                     Driver package catalog file does not contain a signature, and Code Integrity is enforced.
libwdi:debug [syslog] !!!  sig:                     Driver package failed signature validation. Error = 0xE0000247
libwdi:debug [syslog]      sto:                {DRIVERSTORE IMPORT VALIDATE: exit(0xe0000247)} 17:50:04.778
libwdi:debug [syslog] !!!  sig:                Driver package failed signature verification. Error = 0xE0000247
libwdi:debug [syslog] !!!  sto:                Failed to import driver package into Driver Store. Error = 0xE0000247
libwdi:debug [syslog]      sto:           {Stage Driver Package: exit(0xe0000247)} 17:50:04.783
libwdi:debug [syslog]      sto:      {Setup Import Driver Package - exit (0xe0000247)} 17:50:04.791
libwdi:debug [process_message] switching timeout back to finite
libwdi:debug [installer process] Invalid CAT file signature or operation cancelled by user.
libwdi:debug [process_message] installer process completed
Driver Installation: Cancelled by User
6 devices found.
libwdi:debug [wdi_create_list] Hardware ID: USB\VID_0781&PID_5530&REV_0127
libwdi:debug [wdi_create_list] Compatible ID: USB\Class_08&SubClass_06&Prot_50
libwdi:debug [wdi_create_list] Driver version: 10.0.16299.15
libwdi:debug [wdi_create_list] USBSTOR USB device (1): USB\VID_0781&PID_5530\4C530001050719117560
libwdi:debug [wdi_create_list] Device description: 'Cruzer'
libwdi:debug [wdi_create_list] Hardware ID: USB\VID_0403&PID_6010&REV_0700&MI_00
libwdi:debug [wdi_create_list] Compatible ID: USB\Class_ff&SubClass_ff&Prot_ff
libwdi:debug [wdi_create_list] Driver version: 2.12.28.0
libwdi:debug [wdi_create_list] FTDIBUS USB device (2): USB\VID_0403&PID_6010&MI_00\7&1CE5AEFC&0&0000
libwdi:debug [wdi_create_list] Device description: 'Flyswatter2 (Interface 0)'
libwdi:debug [wdi_create_list] Hardware ID: USB\VID_05C8&PID_0369&REV_0108&MI_00
libwdi:debug [wdi_create_list] Compatible ID: USB\Class_0e&SubClass_03&Prot_00
libwdi:debug [wdi_create_list] Driver version: 3.5.8.14
libwdi:debug [wdi_create_list] SPUVCbv USB device (3): USB\VID_05C8&PID_0369&MI_00\7&5B0FA56&0&0000
libwdi:debug [wdi_create_list] Device description: 'HP HD Webcam (Interface 0)'
libwdi:debug [wdi_create_list] Hardware ID: USB\VID_0403&PID_6010&REV_0700&MI_01
libwdi:debug [wdi_create_list] Compatible ID: USB\Class_ff&SubClass_ff&Prot_ff
libwdi:debug [wdi_create_list] Driver version: 2.12.28.0
libwdi:debug [wdi_create_list] FTDIBUS USB device (5): USB\VID_0403&PID_6010&MI_01\7&1CE5AEFC&0&0001
libwdi:debug [wdi_create_list] Device description: 'Flyswatter2 (Interface 1)'
libwdi:debug [wdi_create_list] Hardware ID: USB\VID_8087&PID_07DA&REV_7869
libwdi:debug [wdi_create_list] Compatible ID: USB\Class_E0&SubClass_01&Prot_01
libwdi:debug [wdi_create_list] Driver version: 18.1.1525.1445
libwdi:debug [wdi_create_list] BTHUSB USB device (6): USB\VID_8087&PID_07DA\6&A2946AD&0&6
libwdi:debug [wdi_create_list] Device description: 'Intel(R) Wireless Bluetooth(R)'
libwdi:debug [wdi_create_list] Hardware ID: USB\VID_138A&PID_003F&REV_0104
libwdi:debug [wdi_create_list] Compatible ID: USB\Class_FF&SubClass_00&Prot_00
libwdi:debug [wdi_create_list] Driver version: 4.5.338.0
libwdi:debug [wdi_create_list] WUDFRd USB device (12): USB\VID_138A&PID_003F\00303448B894
libwdi:debug [wdi_create_list] Device description: 'Synaptics FP Sensors (WBF) (PID=003f)'
```

**Comment by pbatard** [comment]
Can you please provide details on how exactly you uninstalled the driver? Did you use Device Manager? Something else? Did you select "Remove the driver files"? Did you reboot your machine before trying to install the driver again?

I am getting sporadic reports for "The requested key container is corrupted", for which I have recently applied a workaround (but that workaround is not yet in a Zadig release), but, even as I tried to replicate the issue, I have never been able to see it on my machine, no matter how many uninstall/reinstall of a driver I go through. I am currently suspecting that there's a component to this that has to do with devices that get a third party driver automatically installed by Windows, which I assume the Flyswatter is (I'm afraid I don't have such a device to test with).

As to your question, I'm afraid that, since I have never experienced this issue, I have no idea how one can manually remove/reset the "corrupted" key container that Windows is complaining about. You may try to use `certmgr`, go to _Trusted Publisher_ and try to remove the `(libwdi autogenerated)` certificates you find there to see if that helps.

If not, I'll see if I can produce a pre-release of Zadig, with the workaround mentioned above, as I'm actually curious to find out if it can provide a solution for this issue.

**Comment by jrhug** [comment]
Thank You for the prompt response.  
When the driver was removed, I used device manager and selected the checkbox to "Delete the driver software for this device".   The system has been rebooted several times.  The system is running Windows 10.  The certificates were removed and the issue is still present.

General information about the Flyswatter2, which may or may not be useful.  The external device has two USB ports.  I am only trying to update one of the ports.  Could this be a problem? 

![image](https://user-images.githubusercontent.com/51538980/106776265-852d2800-6609-11eb-8c38-40fdca38d089.png)

![image](https://user-images.githubusercontent.com/51538980/106775958-3aabab80-6609-11eb-8eb8-d2b8d9d2faf2.png)
